mobile_verifier/src/boosting_oracles/data_sets.rs

```diff
@@ -0,0 +1,482 @@
+        store: FileStore,
+        oracle_boosting_sink: FileSinkClient,
+        data_set_directory: PathBuf,
+    ) -> anyhow::Result<Self> {
+        let latest_file_date = db::fetch_latest_file_date(&pool, T::TYPE).await?;
+        Ok(Self {
+            pool,
+            data_set,
+            data_sets,
+            store,
+            oracle_boosting_sink,
+            data_set_directory,
+            latest_file_date,
+        })
+    }
+
+    fn get_data_set_path(&self, time_to_use: DateTime<Utc>) -> PathBuf {
+        let path = PathBuf::from(format!(
+            "{}.{}.res10.h3tree",
+            T::TYPE.to_prefix(),
+            time_to_use.timestamp()
+        ));
+        let mut dir = self.data_set_directory.clone();
+        dir.push(path);
+        dir
+    }
+
+    pub async fn check_for_available_data_sets(&mut self) -> anyhow::Result<()> {
+        let mut new_data_sets = self
+            .store
+            .list(T::TYPE.to_prefix(), self.latest_file_date, None);
+        while let Some(new_data_set) = new_data_sets.next().await.transpose()? {
+            tracing::info!(
+                "Found new data set: {}, {:#?}",
+                new_data_set.key,
+                new_data_set
+            );
+            db::insert_new_data_set(
+                &self.pool,
+                &new_data_set.key,
+                T::TYPE,
+                new_data_set.timestamp,
+            )
+            .await?;
+            self.latest_file_date = Some(new_data_set.timestamp);
+        }
+        Ok(())
+    }
+
+    pub async fn process_data_sets(&self) -> anyhow::Result<()> {
+        tracing::info!("Checking for new data sets");
+        let mut data_set = self.data_set.lock().await;
+        let latest_unprocessed_data_set =
+            db::fetch_latest_unprocessed_data_set(&self.pool, T::TYPE, data_set.timestamp())
+                .await?;
+
+        let Some(latest_unprocessed_data_set) = latest_unprocessed_data_set else {
+            return Ok(());
+        };
+
+        // If there is an unprocessed data set, download it (if we need to) and process it.
+
+        let path = self.get_data_set_path(latest_unprocessed_data_set.time_to_use);
+
+        // Download the file if it hasn't been downloaded already:
+        if !latest_unprocessed_data_set.status.is_downloaded() {
+            download_data_set(&self.store, &latest_unprocessed_data_set.filename, &path).await?;
+            db::set_data_set_status(
+                &self.pool,
+                &latest_unprocessed_data_set.filename,
+                DataSetStatus::Downloaded,
+            )
+            .await?;
+            tracing::info!(
+                data_set = latest_unprocessed_data_set.filename,
+                "Data set download complete"
+            );
+        }
+
+        // Now that we've downloaded the file, load it into the data set
+        data_set.update(Path::new(&path), latest_unprocessed_data_set.time_to_use)?;
```

Comment:
Should we be checking `time_to_use` here? We could have downloaded a file that's not meant to go live yet.